STM32 ADC: Add new properties to simplify driver

gautierg-st opened the pull request:
Add a set of new properties to the STM32 ADC binding to simplify the driver. These properties describes the HW characteristics of the ADCs (which can sometimes be different for 2 distinct ADCs on the same SoC). This allows the driver to rely on these properties to execute the appropriate functions, instead of discriminating by series/SoC. 

This change improve code readability, and should simplify the addition of new series as there should be less modifications to do inside the driver.

gautierg-st committed:
dts: bindings: adc: stm32: add new properties to simplify the driver

Add a bunch of new property for the STM32 ADC in order to simplify the
driver. All these properties are hardware-specific and should not be
modified by users.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
dts: arm: st: fill out adc nodes with the new properties

Fill out all ADC nodes of all STM32 with the new properties that apply to
them.
Also moves the status at the end of each node.
Also fixes ADC2 and 3 nodes for STM32F103 and ADC3 node for STM32L471 that
were missing some required properties.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: stm32: relies on a ll define rather than series names

For managing the CCRDY flag, rely on the presence of the LL constant
LL_ADC_FLAG_CCRDY rather than a list of series.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: stm32: use the new internal regulator property

Use the new internal regulator property instead of relying on series name
to determine if the regulator should be enabled, and how to check that it
is ready.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: stm32: use the new deep powerdown property

Use the new deep powerdown property instead of relying on the series name
to determine if the ADC needs to be be put out or into deep powerdown mode.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: stm32: use the new channel preselection property

Use the new channel preselection property instead of relying on the series
name to determine if the ADC channels need to be preselecting.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: stm32: use the new differential support property

Use the new differential support property instead of relying on the series
name to determine if the ADC supports differential input channels.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

mathieuchopstm commented on drivers/adc/adc_stm32.c:
```diff
@@ -1646,55 +1661,38 @@ static int adc_stm32_init(const struct device *dev)
+	}
 #endif
 
+#if !ANY_ADC_INTERNAL_REGULATOR_TYPE_IS(INTERNAL_REGULATOR_NONE)
```
This is probably incorrect: `NOT(ANY(xyz))` is not equivalent to `ANY(NOT(xyz))`.
Did you mean:
```suggestion
#if ANY_ADC_INTERNAL_REGULATOR_TYPE_IS(INTERNAL_REGULATOR_DELAY) \
    || ANY_ADC_INTERNAL_REGULATOR_TYPE_IS(INTERNAL_REGULATOR_FLAG)
    /* at least one instance has a regulator */
```

Ditto everywhere where you have this check. You may want to use an intermediary macro e.g. `NEED_INTERNAL_REGULATOR_SUPPORT`

gautierg-st committed:
drivers: adc: stm32: remove useless check of internal regulator state

Disabling the internal regulator is immediate so there is no need to check
the state of the Enable bit in the register.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

etienne-lms reviewed (dismissed):
LGTM. Minor comment to address or not.
